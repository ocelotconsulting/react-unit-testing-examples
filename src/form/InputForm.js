import React, { PureComponent } from 'react'
import T from 'prop-types'

const isEmpty = value => value.trim().length === 0

export default class InputForm extends PureComponent {
  static propTypes = {
    name: T.string.isRequired,
    address: T.string.isRequired,
    onChange: T.func.isRequired,
    onSubmit: T.func.isRequired
  }

  // using a bound instance function has two advantages over inline functions
  // 1 - it's a bit easier to test - we can simplify verify the props of the bound component then test the handler
  // separately
  // 2 - it is theoretically more efficient - when React re-renders an input field it won't think the onChange handler
  // has changed as it would if we inline the function

  onNameChanged = event => {
    this.props.onChange({ name: event.target.value })
  }

  onAddressChanged = event => {
    this.props.onChange({ address: event.target.value })
  }

  render () {
    const { name, address, onSubmit } = this.props
    return (
      <form>
        <div>
          <label htmlFor='registration-form-name'>Name: </label>
          <input type='text' id='registration-form-name' value={name} onChange={this.onNameChanged}/>
        </div>
        <div>
          <label htmlFor='registration-form-address'>Mailing Address: </label>
          <input type='text' id='registration-form-address' value={address} onChange={this.onAddressChanged}/>
        </div>
        <div>
          <button type='button' disabled={isEmpty(name) || isEmpty(address)} onClick={onSubmit}>
            Submit
          </button>
        </div>
      </form>
    )
  }
}
