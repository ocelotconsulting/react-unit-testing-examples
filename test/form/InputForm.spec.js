import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import InputForm from '../../src/form/InputForm'

describe('InputForm', () => {
  let props

  beforeEach(() => {
    // init props with a happy path - that way we can mutate before rendering to set up our test
    props = {
      name: 'Marge Simpson',
      address: '123 Fake Street',
      onChange: jest.fn(),
      onSubmit: jest.fn()
    }
  })

  const shallowRender = () => shallow(<InputForm {...props}/>)

  const getBoundProperties = wrapper => {
    const instance = wrapper.instance()
    return [
      { property: 'name', onChange: instance.onNameChanged },
      { property: 'address', onChange: instance.onAddressChanged }
    ]
  }

  it('binds to fields', () => {
    const wrapper = shallowRender()
    const fieldProps = wrapper.find('input').map(input => input.props())

    expect(fieldProps).toEqual(getBoundProperties(wrapper).map(({ property, onChange }) => ({
      id: `registration-form-${property}`,
      type: 'text',
      value: props[property],
      onChange
    })))
  })

  it('triggers onChange from onChange handlers', () => {
    const wrapper = shallowRender()
    const properties = getBoundProperties(wrapper)

    properties.forEach(({ property, onChange }) => {
      const value = `${property} value`
      onChange({ target: { value } })
      expect(props.onChange).lastCalledWith({ [property]: value })
    })
  })

  it('binds submit button onClick handler', () => {
    expect(shallowRender().find('button').prop('onClick')).toBe(props.onSubmit)
  })

  const isSubmitEnabled = () => !shallowRender().find('button').prop('disabled')

  it('disables submit button when name is empty', () => {
    props.name = ' '

    expect(isSubmitEnabled()).toBe(false)
  })

  it('disables submit button when address is empty', () => {
    props.address = ''

    expect(isSubmitEnabled()).toBe(false)
  })

  it('enables submit button when name and address are non-empty', () => {
    expect(isSubmitEnabled()).toBe(true)
  })

  it('matches snapshot', () => {
    // capture the structure of the rendered component for regression testing
    expect(toJson(shallowRender())).toMatchSnapshot()
  })
})
