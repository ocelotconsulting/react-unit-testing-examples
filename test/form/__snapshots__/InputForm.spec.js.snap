// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputForm matches snapshot 1`] = `
<form>
  <div>
    <label
      htmlFor="registration-form-name"
    >
      Name: 
    </label>
    <input
      id="registration-form-name"
      onChange={[Function]}
      type="text"
      value="Marge Simpson"
    />
  </div>
  <div>
    <label
      htmlFor="registration-form-address"
    >
      Mailing Address: 
    </label>
    <input
      id="registration-form-address"
      onChange={[Function]}
      type="text"
      value="123 Fake Street"
    />
  </div>
  <div>
    <button
      disabled={false}
      onClick={[MockFunction]}
      type="button"
    >
      Submit
    </button>
  </div>
</form>
`;
